### ISSUE TITLE
[Feat] Fix compiler warnings

### ISSUE THREAD

**Issue by nilason** [OSGeo/grass#1247]
The number of compiler warnings are very high, with gcc there are about 140 and with clang about 300 warnings. Most are perhaps trivial and just annoyance, but who knows...
They not only add a lot of noise to build log (you might add to these some 200+ OpenGL deprecation warnings for Mac, and the ctypes warnings addressed in #507), they also makes it very difficult to discover any new potential problems.

**Describe the solution you'd like**
The best would be to address the warnings in the source.

**Describe alternatives you've considered**
It is possible, but in general not advisable, to suppress the warnings with compiler flags.

**Additional context**

*Warning Type	(No. of Warnings)*
*Apple clang version 12.0.0 (clang-1200.0.32.21)*
- [ ] -W#warnings	(16)
- [ ] -Wabsolute-value	(41)
- [ ] -Warray-bounds	(2)
- [ ] -Wc++11-compat-deprecated-writable-strings	(4)
- [ ] -Wdangling-else	(1)
- [ ] -Wdeprecated-declarations	(2)
- [ ] -Wformat	(83)
- [ ] -Wformat-extra-args	(2)
- [ ] -Wformat-invalid-specifier	(4)
- [ ] -Wheader-guard	(8)
- [ ] -Wincompatible-pointer-types	(1)
- [ ] -Wincompatible-pointer-types-discards-qualifiers	(15)
- [ ] -Wlogical-not-parentheses	(5)
- [ ] -Wmacro-redefined	(3)
- [ ] -Wnon-literal-null-conversion	(7)
- [ ] -Wparentheses	(2)
- [ ] -Wparentheses-equality	(3)
- [ ] -Wpointer-bool-conversion	(51)
- [ ] -Wpointer-compare	(4)
- [ ] -Wpointer-sign	(19)
- [ ] -Wswitch	(1)
- [ ] -Wtautological-compare	(1)
- [ ] -Wtautological-pointer-compare	(1)
- [ ] -Wundefined-inline	(16)
- [ ] -Wunused-result	(8)
- [ ] -Wwritable-strings	(5)

**Comment by HuidaeCho** [comment]
I don't think we want to suppress *any* warnings. Ideally, yes, we should fix all of them in the source code. I think we can tackle them one by one (a PR per warning type).

**Comment by nilason** [comment]
> I don't think we want to suppress _any_ warnings. Ideally, yes, we should fix all of them in the source code. I think we can tackle them one by one (a PR per warning type).

As you have seen, this is my intention. But not PR per warning type, but grouped with a commit per warning type which will make it all (I think) more handy.

I’m looking forward to your input!

**Comment by HuidaeCho** [comment]
> > I don't think we want to suppress _any_ warnings. Ideally, yes, we should fix all of them in the source code. I think we can tackle them one by one (a PR per warning type).
> 
> As you have seen, this is my intention. But not PR per warning type, but grouped with a commit per warning type which will make it all (I think) more handy.
> 
> I’m looking forward to your input!

So you mean one commit per warning type, but all these commits in one PR? That could also work, but we have to think about a potential need for reversing just "some" of those commits for whatever reasons. All warnings per module in one PR is also an option. This way, if something is broken in one module because of these fixes, we can easily go back and inspect just one module.

**Comment by nilason** [comment]
The PRs "Fix compiler warnings" part 1–12, addresses all above listed warnings for Clang (Apple) and gcc (Linux).
In other words, the 12th PR is the last (planned) PR in this campaign.

**Comment by nilason** [comment]
There are 2 still open PR's addressing this issue with compiler warnings. Both are, I believe, close to be merged. Would be good, although not necessary, to have these changes with G8.

The PR's, when merged, will reduce the number of compiler warnings of the CI's "GCC C/C++ standards check" to ZERO! I will then be possible, if so wished, to trigger a CI failure with newly introduced warnings.

**Comment by nilason** [comment]
> There are 2 still open PR's addressing this issue with compiler warnings. Both are, I believe, close to be merged. Would be good, although not necessary, to have these changes with G8.
> 
> The PR's, when merged, will reduce the number of compiler warnings of the CI's "GCC C/C++ standards check" to ZERO! I will then be possible, if so wished, to trigger a CI failure with newly introduced warnings.

With #1804 the GCC CI builds fails with compiler warnings.

**Comment by wenzeslaus** [comment]
> ...I will then be possible, if so wished, to trigger a CI failure with newly introduced warnings.

I think that's the only way to keep it at zero. You can add a clang or macOS (presumably with clang) workflow which will check specifically that (at least in the past in C++ there were differences between GCC and clang in what was considered error, not even talking about warnings). A clang check could be potentially integrated in the gcc one using the keys in `matrix.include`, but whatever is easier works.
